internal/settings/providers/dynu/provider.go

```diff
@@ -0,0 +1,155 @@
+package dynu
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/settings/constants"
+	"github.com/qdm12/ddns-updater/internal/settings/errors"
+	"github.com/qdm12/ddns-updater/internal/settings/headers"
+	"github.com/qdm12/ddns-updater/internal/settings/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+)
+
+type provider struct {
+	domain        string
+	host          string
+	ipVersion     ipversion.IPVersion
+	username      string
+	password      string
+	useProviderIP bool
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *provider, err error) {
+	extraSettings := struct {
+		Username      string `json:"username"`
+		Password      string `json:"password"`
+		UseProviderIP bool   `json:"provider_ip"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	p = &provider{
+		domain:        domain,
+		host:          host,
+		ipVersion:     ipVersion,
+		username:      extraSettings.Username,
+		password:      extraSettings.Password,
+		useProviderIP: extraSettings.UseProviderIP,
+	}
+	if err := p.isValid(); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *provider) isValid() error {
+	switch {
+	case len(p.username) == 0:
+		return errors.ErrEmptyUsername
+	case len(p.password) == 0:
+		return errors.ErrEmptyPassword
+	case p.host == "*":
+		return errors.ErrHostWildcard
```

Comment:
Wildcards are not supported then right? Are custom hosts (i.e. `a.example.com`) supported?

Reply:
If I use @ or * it updates all DDNS IPs for my domains on Dynu, * doesn't update subdomains of a concrete domain, so `*.domain.com` updates domain.com but not `subdomain.domain.com` 

There's an `alias` (subdomain) parameter that could be used to update a concrete subdomain, to update a bunch of domains and subdomains in one call there's an optional parameter called `location` which could be used if configured in Dynu, would be similar to * but you have to set your domains and subdomains to a group in dynu.

I'm adding more features to support all these parameters, I'll update the pull as soon as I finish.

Reply:
> If I use @ or * it updates all DDNS IPs for my domains on Dynu, * doesn't update subdomains of a concrete domain, so *.domain.com updates domain.com but not subdomain.domain.com

I'm a bit confused here.
Correct me if I'm wrong:
- `@` for domain.com updates only `domain.com` but not `sub.domain.com
- `*` for domain.com updates `domain.com`, `domain2.com` but not `sub.domain.com`

Reply:
Well, `@` does nothing really, if you don't set the `hostname` or if it's not a valid hostname for your user it will update all records for your domains that don't have a group assigned.

`*` doesn't acts like a wildcard, you can use it in the hostname but it will update only the `domain.com` and not `sub.domain.com`